launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java

```diff
@@ -95,10 +95,12 @@
 
     if (javaHome != null) {
       cmd.add(join(File.separator, javaHome, "bin", "java"));
+    } else if (childEnv.containsKey("JAVA_HOME")) {
```

Comment:
I think this change could be fine, but I wonder if we can rewrite the logic so that these cases update a variable tracking what the Java home is, and add the command once at the end. The logic is duplicated four times here.

Reply:
@srowen The logic has been updated, please review the code, thanks.